markets/storageadapter/dealpublisher.go

```diff
@@ -443,3 +467,63 @@ func (p *DealPublisher) filterCancelledDeals() {
 	}
 	p.pending = filtered
 }
+
+func (p *DealPublisher) PublishQueuedDeals(deals []cid.Cid) []cid.Cid {
+	p.lk.Lock()
+	var ret []cid.Cid
+	var toPublish []*pendingDeal
+
+	// Check that each deal is part of the queue
+	for _, c := range deals {
+		found := false
+		for _, pd := range p.pending {
+			if c.Equals(pd.proposalCid) {
+				found = true
+				toPublish = append(toPublish, pd)
+				break
+			}
+		}
+		if found {
+			ret = append(ret, c)
+		} else {
+			log.Debugf("failed to find the proposal %S in pending deals", c)
+		}
+	}
+
+	p.lk.Unlock()
+
+	log.Infof("publishing deal proposals: %s", ret)
+
+	// Remove the deal from Pending
+	p.cleanupPending(deals)
```

Comment:
I'd suggest we clean up pending deals under the lock also, so that the whole operation is atomic.
Otherwise two threads could interfere with each other.
Maybe you can even refactor so that it splits the list into the items to be published and the items that will remain (you're doing this in separate for loops at the moment, probably you can do both in the same for loop).

Reply:
I have refactored the code a little. Unique identifier don't go well with slice.
The new code looks more clean. It resulted in some minor changes in other functions as well.